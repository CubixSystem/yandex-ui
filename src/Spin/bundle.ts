import { compose, composeU, ExtractProps } from '@bem-react/core';

import { Spin as SpinCommon } from './Spin';
// _position
import { withPositionCenter } from './_position/Spin_position_center';
//  _size
import { withSizeL } from './_size/Spin_size_l';
import { withSizeM } from './_size/Spin_size_m';
import { withSizeS } from './_size/Spin_size_s';
import { withSizeXS } from './_size/Spin_size_xs';
import { withSizeXXS } from './_size/Spin_size_xxs';
// _view
import { withViewDefault } from './_view/Spin_view_default';

export * from './Spin';

export const Spin = compose(
    composeU(withSizeL, withSizeM, withSizeS, withSizeXS, withSizeXXS),
    withPositionCenter,
    withViewDefault,
)(SpinCommon);

export type ISpinProps = ExtractProps<typeof Spin>;
