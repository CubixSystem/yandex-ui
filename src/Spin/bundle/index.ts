export * from '../Spin.bundle/common';
