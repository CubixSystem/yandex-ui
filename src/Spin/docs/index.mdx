---
id: Spin
path: /components/spin/examples

title: Spin
description: Индикатор загрузки. Отображает выполнение какого-то процесса, например загрузки сайта или медиафайла.

links:
 - label: source
   url: https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/Spin
 - label: startrek
   url: https://search.yandex-team.ru/stsearch?text=Spin.ts&facet.queue=ISL&facet.type=bug&facet.status=128
 - label: figma
   url: https://www.figma.com/file/CaoHFEO0w6PaE8cVaYFDoq8i/L-E-G-O?node-id=0%3A1
 - label: storybook
   url: https://lego.yandex-team.ru/components/lego/latest/?storybook=%2Fstory%2Fprogress-spin-desktop--playground

tabs:
 - Examples
 - Code
---
