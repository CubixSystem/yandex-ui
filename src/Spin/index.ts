export * from './Spin';
export * from './_position/Spin_position_center';
export * from './_size/Spin_size_l';
export * from './_size/Spin_size_m';
export * from './_size/Spin_size_s';
export * from './_size/Spin_size_xs';
export * from './_size/Spin_size_xxs';
export * from './_view/Spin_view_default';
