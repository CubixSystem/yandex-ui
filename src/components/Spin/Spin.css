.Spin2 {
    display: none;

    box-sizing: border-box;

    border: 2px solid transparent;
    border-radius: 50%;
    background-image: linear-gradient(to right, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0));
}

/* Для правильного позиционирования блока относительно baseline */
.Spin2::after {
    visibility: hidden;

    content: '\00A0'; /* &nbsp; */
}

/*
  (1) https://st.yandex-team.ru/ISL-6342
      Добавлена задержка начала анимации, что заставляет её нормально работать в iOS 12-
      Решение взято отсюда:
      https://stackoverflow.com/questions/29219534/css-animation-not-always-starting-in-ios-8-safari
*/
.Spin2_progress {
    display: inline-block;

    animation: islands-spin2 1s .21s infinite linear; /* (1) */

    backface-visibility: hidden;  /* Для ускорения анимации */
}

@keyframes islands-spin2 {
    from {
        transform: rotate(0deg);
    }

    to {
        transform: rotate(360deg);
    }
}
