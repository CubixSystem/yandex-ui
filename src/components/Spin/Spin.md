# Spin

<a href='https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/components/Spin' target='_blank'><img src='https://badger.yandex-team.ru/custom/[Исходники]/[Github
][green]/badge.svg' /></a> <a href='https://search.yandex-team.ru/stsearch?text=Spin.ts&facet.queue=ISL&facet.type=bug&facet.status=128' target='_blank'><img src='https://badger.yandex-team.ru/custom/[Известные проблемы]/[Startrek][blue]/badge.svg' /></a>

<!-- description:start -->
Индикатор загрузки. Отображает выполнение какого-то процесса, например загрузки сайта или медиа-файла.
<!-- description:end -->

## Пример использования

Конфигурация темы на уровне проекта:

```ts
// src/lib/theme.ts
import { configureRootTheme } from '@yandex-lego/components/Theme'
import { theme } from '@yandex-lego/components/Theme/presets/default'

configureRootTheme({ theme })
```

Использование с нужным набором модификаторов:

```ts
// src/App.ts
import React from 'react'
import { compose } from '@bem-react/core'
import {
  Spin as SpinDesktop,
  withSizeM,
  withViewDefault,
} from '@yandex-lego/components/Spin/desktop'

// Композиция из различных модификаторов
const Spin = compose(withSizeM, withViewDefault)(SpinDesktop)

const App = () => {
  return <Spin progress view="default" size="m" />
}
```

Использование с полным набором модификаторов:

```ts
// src/App.ts
import React from 'react'
import { Spin } from '@yandex-lego/components/Spin/desktop/bundle'

const App = () => {
  return <Spin progress view="default" size="m" />
}
```

## Примеры

### Размер индикатора

Чтобы изменить размер индикатора, установите свойство `size` в одно из следующих значений: `"l"`, `"m"`, `"s"`, `"xs"`, `"xxs"`.

{{%story::desktop:progress-spin-desktop--size%}}

### Расположение индикатора

Чтобы отобразить индикатор по центру, установите свойство `position` в значение `center`.

{{%story::desktop:progress-spin-desktop--position%}}

## Свойства

<!-- props:start -->
| Свойство   | Тип                         | Описание                                   |
| ---------- | --------------------------- | ------------------------------------------ |
| innerRef?  | `RefObject<HTMLDivElement>` | Ссылка на корневой DOM-элемент компонента. |
| progress?  | `false \| true`             | Видимость индикатора загрузки.             |
| className? | `string`                    | Дополнительный класс.                      |
<!-- props:end -->
