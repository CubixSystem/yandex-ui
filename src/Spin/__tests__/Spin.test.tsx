import React, { createRef } from 'react';

import { createClientRender, screen } from '../../internal/testing';
import { Spin } from '../Spin';

describe('Spin', () => {
    const render = createClientRender();

    test('должен вернуть полный шаблон компонента (snapshot)', () => {
        const { container } = render(<Spin className="mix" />);

        expect(container.firstChild).toMatchSnapshot();
    });

    test('должен быть установлен корректный displayName', () => {
        expect(Spin.displayName).toBe('Spin2');
    });

    test('должен устанавливать ссылку на корневой DOM элемент', async () => {
        const innerRef = createRef<HTMLDivElement>();
        render(<Spin innerRef={innerRef} />);

        expect(innerRef.current).not.toBe(null);
    });

    test('должен устанавливать/удалять модификатор progress при обновлении через props', () => {
        const { setProps } = render(<Spin data-testid="element" />);

        expect(screen.getByTestId('element')).not.toHaveClass('Spin2_progress');

        setProps({ progress: true });

        expect(screen.getByTestId('element')).toHaveClass('Spin2_progress');
    });
});
