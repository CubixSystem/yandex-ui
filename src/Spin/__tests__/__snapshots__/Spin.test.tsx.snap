// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spin должен вернуть полный шаблон компонента (snapshot) 1`] = `
<div
  class="Spin2 mix"
/>
`;
